### PR TITLE
fix: improve RFC 9110 compliance across req/res/ctx, helpers, and middleware

### PR DESCRIPTION
**Pull request by gaby**
Audit of req.go, res.go, ctx.go, and their shared helpers against RFC 9110
(HTTP Semantics), plus the adaptor and proxy middlewares, uncovered a set of
compliance gaps. This PR fixes them together with regression tests,
regenerated interface docs, and updated docs/api/ctx.md.

## Request parsing (req.go)

- Fresh(): only report freshness for GET and HEAD. A 304 Not Modified is
  defined only for those methods, and Section 13.1.3 requires If-Modified-Since
  to be ignored otherwise.
- Fresh(): accept the obsolete RFC 850 and ANSI C asctime() HTTP-date formats
  in addition to IMF-fixdate (Section 5.6.7), falling back to net/http's
  ParseTime for the obsolete forms.
- Charset(): parse Content-Type parameters as quoted-strings, so a ";" or a
  stray quote inside a value no longer breaks charset extraction
  (Sections 5.6.6 and 5.6.4).
- Method(): match the override case-sensitively first (Section 9.1) so
  mixed-case custom methods registered via Config.RequestMethods work, then
  fall back to the uppercase form for the standard methods. Unregistered
  methods are reported from the raw header instead of panicking.
- Range(): distinguish an unsupported range unit (which Section 14.2 requires
  the server to ignore, now ErrRangeUnsupported) from a malformed header
  (Section 14.1.1); parse and ignore empty list elements (Section 5.6.1.2);
  reject a last-byte-pos below its first-byte-pos.
- Body(), AcceptEncoding(), AcceptLanguage(): combine repeated field lines into
  one comma-joined list (Section 5.2), matching what the negotiation helpers
  operate on.

## Response building (res.go)

- Vary(): treat "*" per Section 12.5.5, where the wildcard is only meaningful
  as the sole member; adding "*" collapses the header to a single "*".
- Format(): with no Accept header, skip "default" entries when selecting the
  handler so the literal string "default" is never emitted as a Content-Type.
- AutoFormat(): mark the response as Vary: Accept, since the representation is
  chosen by proactive negotiation (Section 12.5.5).
- Attachment()/Download(): emit an RFC 6266 quoted-string filename with
  RFC 9110 Section 5.6.4 escaping, and an RFC 8187 filename* ext-value with
  correct attr-char percent-encoding (url.PathEscape left '=', '@', ':' bare).
- Append(): skip empty list elements (Section 5.6.1.2) and dedup across all
  combined field lines (Section 5.2).
- Links(): escape the rel value inside its quoted-string (Section 5.6.4).
- SendEarlyHints(): only send the interim 103 on HTTP/1.1 connections
  (Section 15.2); the Link headers still go out on the final response.

## Shared helpers (ctx.go, helpers.go, error.go, app.go)

- IsWebSocket(): treat Connection and Upgrade as combined lists (Section 5.2)
  and accept an Upgrade protocol with a "/version" suffix (Section 7.8).
- hasTransferEncodingBody(): inspect every Transfer-Encoding field line
  (Section 5.2).
- isEtagStale(): split If-None-Match on commas outside DQUOTE, so a "," inside
  a quoted opaque-tag stays part of one entity tag (Section 8.8.3).
- getOffer(): match the weight parameter name "q" case-insensitively
  (Section 12.4.2).
- getSplicedStrList(): drop empty list elements (Section 5.6.1.2).
- Range errors now carry a 400 Bad Request status so blind propagation does not
  surface as a 500; removed the now-unused toBytes/quoteString path and range
  sentinels.

## Middleware

- adaptor: propagate the request's real protocol version, and initialize the
  fasthttp RequestCtx with a no-op connection so an interim response (e.g. a
  103 from SendEarlyHints) is discarded gracefully instead of panicking; the
  final response still reaches the client through the ResponseWriter copy-back.
- proxy: reset the protocol token to HTTP/1.1 before forwarding, so a version
  inherited from the inbound request (e.g. "HTTP/2" from the adaptor) does not
  produce an invalid upstream request line.

## Notable behavior changes

- Content-Disposition filenames are now literal-in-quotes (e.g.
  filename="report 2024.txt" instead of "report+2024.txt").
- Fresh() returns false for methods other than GET and HEAD.
- Method() overrides are matched case-sensitively before the uppercase fallback.

Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>
Claude-Session: https://claude.ai/code/session_01FST7xR3C6zqK5wWzBCv4c5

### CODE REVIEW

File: ctx.go [hunk 1]
```diff
@@ -453,26 +454,33 @@ func (c *DefaultCtx) OverrideParam(name, value string) {
...
+			continue
+		}
+		if transferEncodingLineHasBody(utils.UnsafeString(value)) {
+			return true
+		}
 	}
 
-	hasEncoding := false
+	return false
+}
```
[the remark this answers is not in the document]
Comment (gaby) [reply ctx.go:484]: The suggested isolation is already in place as far as it can be detected: `PeekAll` runs first, and the `hdr.All()` fallback only executes when the normalized lookup returned nothing. The edge case it exists for — header keys stored non-normalized (e.g. `DisableHeaderNamesNormalizing` / `DisableSpecialHeader` configurations, which the tests exercise) — is indistinguishable from "header absent" without scanning, so the fallback can't be gated further without dropping that correctness.

The cost is also smaller than it looks: `utils.EqualFold` short-circuits on length first, so the fallback walk is N integer comparisons for a typical header set, and `hasTransferEncodingBody` only runs at all when `Content-Length == 0` inside an explicit `HasBody()` call — it's not on the framework's per-request path. This structure (Peek fast path + `All()` fallback) predates this PR; the change here only made the fallback inspect every matching field line instead of just the first, per RFC 9110 §5.2.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by gaby** [comment]
Benchmarks have been fixed, the comment didn't get updated by the workflow.

**Review by ReneWerner87 — APPROVED** [review]
